### DISCUSSION TITLE
AG itself creates panic

### DISCUSSION THREAD

**Discussion by manatactical** [apohl79/audiogridder#1038] [Help]
![Screen Shot 2022-09-19 at 10 41 06 PM](https://user-images.githubusercontent.com/9537424/191112945-1fb38eeb-70c8-4608-bdd2-b870691860f3.png)

Happens on Logic and Cubase, with beta11 and beta13. I have no idea why. (Setting the DAW buffer to 2048 doesn't help either).
Dark magic is happening. Any idea?

Loading a VST3 instrument makes things better, VST2 seems to be pretty unstable tho.

PS: RME Drivers are rock-solid. Usually bulletproof when it comes to ASIO load.

**Comment by apohl79** [comment]
Beta13 has more metrics in the monitor and faster performance updates. That would be the starting point. You can also try to disable "Use fixed outbound buffers". Also you can enable the "RealTime Mode" in b13 under "Transfer Audio/MIDI". This should avoid (or reduce) the peaks in the DAW but will probably show read errors, if the data does not come back fast enough.

**Reply by manatactical** [reply]
Thanks! I'm going to try those settings.

**Reply by manatactical** [reply]
Realtime with 0 blocks does help. However, using a 256 buffer (which is the one I use) causes a huge ASIO load.

![Screen Shot 2022-09-20 at 11 14 16 AM](https://user-images.githubusercontent.com/9537424/191219149-f2cd7b0e-8372-444b-a09f-fdcaa5785aac.png)
![Screen Shot 2022-09-20 at 11 14 33 AM](https://user-images.githubusercontent.com/9537424/191219161-286d4517-ec97-403a-b4f0-00f9f489fbde.png)
![Screen Shot 2022-09-20 at 11 14 49 AM](https://user-images.githubusercontent.com/9537424/191219166-fb6a136f-d6a4-4519-adbd-dabc6529f358.png)

With 2 more blocks:
![Screen Shot 2022-09-20 at 11 17 40 AM](https://user-images.githubusercontent.com/9537424/191219596-1f61d7a9-d1b9-43cc-a1c6-dfb2dde6aaf5.png)

**Reply by apohl79** [reply]
Did you check the plugin monitor? Usually the load should be reduced when using a buffer, as the processing time decreases. What processing time (Perf Proc) and network transfer time (Perf Net) do you see in the monitor with and without buffering?

**Reply by manatactical** [reply]
Buffer 256 with and without 2 blocks:
![Screen Shot 2022-09-20 at 12 52 56 PM](https://user-images.githubusercontent.com/9537424/191239996-279a7b8c-09f2-4362-8fb7-3f174e9faa10.png)
![Screen Shot 2022-09-20 at 12 53 23 PM](https://user-images.githubusercontent.com/9537424/191240000-f5b0a01b-3770-4ce7-aac8-a813259f7f27.png)

Buffer 128 with and without 2 blocks:
![Screen Shot 2022-09-20 at 12 53 45 PM](https://user-images.githubusercontent.com/9537424/191240060-5fabc81a-b1c5-4238-ad88-3043a6f49788.png)
![Screen Shot 2022-09-20 at 12 54 04 PM](https://user-images.githubusercontent.com/9537424/191240065-4c83e61b-af9e-4fb8-8f5a-2aefabddc1f1.png)

**Reply by manatactical** [reply]
What is ReadError? What it detects exactly?

Sometimes it somehow hangs and the count skyrockets. (I have to reload to fix the issue. Happen at 512 or 128 after some time,but it's triggered instantly at 256)

![Screen Shot 2022-09-20 at 1 19 19 PM](https://user-images.githubusercontent.com/9537424/191244856-57ad32e4-8fc6-4635-b0c9-f8635901fff8.png)

**Reply by apohl79** [reply]
Read Error means AG drops audio as it does not come back fast enough. But the metric on the last screenshot does not look ok as it states the network part takes only 0.7ms while processing itself takes 4ms. This should not happen.

Would you mind to enable tracing in the plugin under diagnostics (restart the DAW afterwards), let a run for a bit in the 2 blocks setting and send over the log/trace files? (https://audiogridder.com/bug-report-diagnostics-locations/)

**Reply by manatactical** [reply]
Sure thing! I'm on it :)

**Reply by apohl79** [reply]
I can't reproduce this unfortunately (tried 256 samples, no asio guard, 2 blocks). And the trace does not contain useful info (it has only very recent data due to the size limit, just a few seconds usually).

Some assumptions:

It looks like the server is not responding fast enough. Looking at your log, there is entries like this:

[14:09:26.640|client|7f8211c423d8] warning: instance (Kontakt): read queue empty, waiting for data, try to increase the buffer
[14:09:26.649|client|7f8211c423d8] error: instance (Kontakt): waitRead failed
[14:09:26.649|client|7f8211c423d8] process block took 9.14032ms (d5e51e13-36c9-4ecd-9b50-c2e022ff332c)
[14:09:26.649|client|7f8211c423d8]   - pb_bypass_chk: 0.002009ms
[14:09:26.649|client|7f8211c423d8]   - pb_prep: 0.01059ms
[14:09:26.649|client|7f8211c423d8]   - pb_get_streamer: 0.000291ms
[14:09:26.649|client|7f8211c423d8]   - pb_ch_map: 0.003256ms
[14:09:26.649|client|7f8211c423d8]     + as_prep: 0.004786ms
[14:09:26.649|client|7f8211c423d8]     + as_copy_to_wbuf: 0.008147ms
[14:09:26.649|client|7f8211c423d8]     + as_upd_pos: 0.000179ms
[14:09:26.649|client|7f8211c423d8]     + as_copy_from_wbuf: 0.016804ms
[14:09:26.649|client|7f8211c423d8]     + as_push: 0.003493ms
[14:09:26.649|client|7f8211c423d8]     + as_notify: 0.002469ms
[14:09:26.649|client|7f8211c423d8]   = pb_send: 0.037301ms
**[14:09:26.650|client|7f8211c423d8]     = as_wait_read_failed: 9.07671ms
[14:09:26.650|client|7f8211c423d8]   = pb_read: 9.07873ms**
[14:09:26.650|client|7f8211c423d8]   - pb_ch_map_reverse: 0.002465ms
[14:09:26.650|client|7f8211c423d8]   - pb_update_midi_state: 0.00023ms
[14:09:26.650|client|7f8211c423d8]   - pb_finish: 0.000147ms

The bold lines tell you that the read timed out after ~9ms. Now you can check the server log for a matching entry for the ID **d5e51e13-36c9-4ecd-9b50-c2e022ff332c** (which is the correlation ID of the audio transaction that timed out). If you find an entry, you might see that Kontakt did take more time than usual or that something else was taking time. You might not find server entries for everything in the plugin log, but hopefully a few. If you don't find any, there might be something else, maybe on the network or so. Really hard to say.

**Reply by manatactical** [reply]
Thank you for checking the log. Very odd indeed. I'm gonna check the log on the server side. AG has literally 4 cores and a 2.5GB Network card for itself. Maybe a network parameter is messing around. But why 256? that's voodoo magic right there haha.

I will keep you posted.

**Reply by manatactical** [reply]
Ok. Got some news already.

Disabling the screen capture seems to cause the issue at 256 (and maybe the hanging after a while). Having it enabled at 256 didn't trigger the load spikes. However, it seems there are still issues in the log related to it.


[AudioGridderServer_2022-09-20_17-37-48.log](https://github.com/apohl79/audiogridder/files/9609382/AudioGridderServer_2022-09-20_17-37-48.log)

**Reply by apohl79** [reply]
Does it get better if you change the sandbox mode to "chain isolation" ?

**Reply by manatactical** [reply]
Nope

**Reply by apohl79** [reply]
Ok I was just able to reproduce a "hang" switching the kontakt UI on and off a bunch of times. Will see if that situation can be improved. Not sure yet if it happens only with the plugin isolation mode. But thats where I saw it happening.

**Reply by apohl79** [reply]
> Nope

ok cheers

**Reply by manatactical** [reply]
Glad you've been able to reproduce it. We are on the right path I guess. Sorry for bringing a weird issue to the table haha.
Thanks again for the great work and support!

**Reply by apohl79** [reply]
I think I sorted it out, should be better in the next beta. That was a good catch, so thx!

**Reply by manatactical** [reply]
Superb! Look forward to trying the new beta out.

**Comment by manatactical** [comment]
Hi there.

Beta15, and as I don't need the screen capture anymore I disabled it, it's the same mess. I have no idea why it behaves like that. Spiking everywhere, audio cut after glitches for a few seconds. It hands and produces hundreds of errors per second. Changing audio settings don't do anything.

`[16:51:32.865|client|7fd66b1701d8] error: instance (SINE Player): write queue full, dropping samples
[16:51:32.865|client|7fd66b1701d8] warning: instance (SINE Player): read queue empty, waiting for data, try to increase the buffer
[16:51:32.869|client|7fd66b1701d8] error: instance (SINE Player): waitRead failed
[16:51:32.869|client|7fd66b1701d8] process block took 4.03375ms (7b032336-0796-4c6a-b3ff-df7ea1fd8424)
[16:51:32.869|client|7fd66b1701d8]   - pb_bypass_chk: 0.000261ms`

**Reply by manatactical** [reply]
It does happen with Kontakt too, or anything.
